File: semester_1/homework_6/stack.cpp
#include "stack.h"

#include <iostream>

struct stackStuff::Stack
{
    int *storage = nullptr;

    int size = 0;
    int capacity = 0;
};

stackStuff::Stack *stackStuff::createStack(int maxSize = DEFAULT_SIZE)
{
    Stack *newStack = new Stack;
    newStack->capacity = maxSize;
    newStack->storage = new int[maxSize];

    return newStack;
}

void stackStuff::deleteStack(Stack *toDeleteStack)
{
    delete[] toDeleteStack->storage;
    delete toDeleteStack;
}

bool stackStuff::isEmpty(const Stack *inputStack)
{
    return (inputStack->size == 0);
}

int stackStuff::top(const Stack *inputStack)
{
    if (isEmpty(inputStack))
    {
        std::cerr << "The stack is empty!!!" << std::endl;
        return -1;
    }

    return inputStack->storage[inputStack->size - 1];
}

void stackStuff::push(Stack *inputStack, int value)
{
    if (inputStack->size == inputStack->capacity)
    {
        std::cerr << "The stack is full!!!" << std::endl;
        return;
    }

    inputStack->storage[inputStack->size] = value;
    inputStack->size++;
}

int stackStuff::pop(Stack *inputStack)
{
    if (isEmpty(inputStack))
    {
        std::cerr << "The stack is empty!!!" << std::endl;
        return -1;
    }

    return inputStack->storage[--inputStack->size];
}

File: semester_2/homework_2/problem_1/Stack.cs
namespace StackStuff
{
    public class Stack<T> : IStack<T>
    {
        private int size = 0;
        private StackElement top = default(StackElement);

        public Stack()
        {
        }

        public bool IsEmpty()
        {
            return this.size == 0;
        }

        public T Pop()
        {
            if (this.IsEmpty())
            {
                // TODO: Replace with exception
                return default(T);
            }
            else
            {
                var returnValue = this.top;

                this.top = this.top.Next;
                --this.size;

                return returnValue.Value;
            }
        }

        public void Push(T value)
        {
            this.top = new StackElement(value, this.top);
            ++this.size;
        }

        private class StackElement
        {
            public T Value;
            public StackElement Next;

            public StackElement(T value, StackElement next)
            {
                this.Value = value;
                this.Next = next;
            }
        }
    }
}